feat(discover): Support release:latest in discover

wmak opened the pull request:
- Adds support for `release:latest` in discover
    - Changed `environment` in `get_filter_params` to be the environment objects
    - Adding organization_id as a default there as well.
- Adds `latest` as an autocomplete option in the searchbar

wmak committed:
feat(discover): Supporting release:latest in queries

- Changed how `environment` is passed via params to make it easier to
  use

wmak committed:
feat(searchbar): Adding latest to autocomplete

- When the tag key is release, this adds `latest` tas a possible value

markstory commented on src/sentry/api/bases/organization.py:
```diff
@@ -276,8 +276,13 @@ def get_filter_params(self, request, organization, date_filter_optional=False):
 
-        environments = [env.name for env in self.get_environments(request, organization)]
-        params = {"start": start, "end": end, "project_id": [p.id for p in projects]}
+        environments = self.get_environments(request, organization)
```
This seems like a risky change. Couldn't this be used in many endpoints?

wmak replied:
oof yeah, this breaks a lot of things. I'll try to find a less risky way to do this :+1:

markstory commented on src/sentry/search/utils.py:
```diff
@@ -224,10 +224,11 @@ def parse_user_value(value, user):
-    )
+def get_latest_release(projects, environments, organization_id=None):
+    if organization_id is None:
+        organization_id = projects[0].organization_id
```
When would you potentially want a different organization_id than what is on the project?

wmak replied:
In this case its because we only have a list of project_ids but still have access to the organization_id

markstory replied:
Should we update the other usages of this to also include the organization_id?

wmak replied:
Other usage of this is in, `GroupEventsEndpoint` which gets the org even for other functions via `group.project.organization`

wmak committed:
ref: Undoing changes to environment param

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -883,7 +902,7 @@ def to_list(value):
                 kwargs["project_ids"] = params["project_id"]
         if "environment" in params:
-            term = SearchFilter(SearchKey("environment"), "=", SearchValue(params["environment"]))
+            term = SearchFilter(SearchKey("environment"), "=", SearchValue(params["environment"]),)
```
```suggestion
            term = SearchFilter(SearchKey("environment"), "=", SearchValue(params["environment"]))
```

evanh commented on src/sentry/api/endpoints/organization_events.py:
```diff
@@ -126,8 +126,6 @@ def get(self, request, organization):
             except NoProjects:
                 return Response([])
 
-            params["organization_id"] = organization.id
```
Why did you remove this? It seems like you're using it when you call `parse_release`

wmak replied:
Aiming for consistency where we set stuff on the `params` dict, I moved this into the `get_filter_params` [here](https://github.com/getsentry/sentry/pull/18568/files#diff-f348b3a3d06653db5b246ec567d39dc4R286)

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -853,6 +855,23 @@ def to_list(value):
+                        SearchValue(
+                            parse_release(
+                                term.value.value,
+                                params["project_id"],
```
This seems like it might break, since `get_latest_release` expects actual project objects, not IDs (`organization_id = projects[0].organization_id`)

wmak replied:
Poor naming in `get_filter_params` `project_id` is a list of objects, not ids see [here](https://github.com/getsentry/sentry/pull/18568/files#diff-f348b3a3d06653db5b246ec567d39dc4L282)

Worth changing in another PR I think though since that was my first impression of that too

wmak replied:
Sorry previous comment was completely incorrect, this is more a django thing
```
Release.objects.filter(project__in=[1,2,3])
```

```
projects = [p for p in Project.objects.filter(id__in=[1,2,3])]
Release.objects.filter(project__in=projects)
```

Become the same query with `...WHERE "sentry_release_project"."project_id" IN (1, 2, 3)`

evanh replied:
Won't this line fail: https://github.com/getsentry/sentry/pull/18568/files#diff-712e897580b290560333dc7b4bb0dd88R229

Since an int does not have an attribute 'organization_id'.

wmak replied:
Hmm I'll add a check there, but in this case the organization_id is passed from `params` so the function will use that instead of trying to pull it off the projects.

The function is shared with `GroupEventsEndpoint` which only has the list of project objects, but no org which is why it has to grab the org_id

wmak committed:
Merge branch 'master' into feat/discover-latest-release

wmak committed:
Merge branch 'master' into feat/discover-latest-release

wmak committed:
fix(discover): Fixing wildcard release searches

- Only checking latest when the term value is `latest` so we don't have
  to repeeat work to make it a wildcard search
- Adding a value var to make it easier to get

wmak committed:
Update src/sentry/api/event_search.py

Co-authored-by: evanh <evanh@users.noreply.github.com>

wmak committed:
fix: Adding a check that project has an organization_id

- Also adding a test for this

wedamija commented on src/sentry/api/event_search.py:
```diff
@@ -829,30 +830,47 @@ def to_list(value):
+                                value,
+                                params["project_id"],
+                                params.get("environment_objects"),
+                                params["organization_id"],
```
@wmak `organization_id` isn't always in params here, and causes this error: https://sentry.io/organizations/sentry/issues/1846529501/?project=1&query=is%3Aunresolved+assigned%3A%23workflow&statsPeriod=14d. Mind taking a look?

wmak replied:
:+1: can do thanks!

wmak replied:
Thanks again for the heads up, fix here: https://github.com/getsentry/sentry/pull/20790
